notebooks/practical_issues/TargetEncoding.ipynb

```diff
@@ -0,0 +1,4786 @@
+    "                                  prediction_task=task)\n",
+    "\n",
+    "# Add our local models\n",
+    "first_model = CertifaiModel('SVM',\n",
+    "                            local_predictor=svm_model_proxy)\n",
+    "scan.add_model(first_model)\n",
+    "second_model = CertifaiModel('logistic',\n",
+    "                            local_predictor=logistic_model_proxy)\n",
+    "scan.add_model(second_model)\n",
+    "\n",
+    "# Add the eval dataset\n",
+    "eval_dataset = CertifaiDataset('evaluation',\n",
+    "                               CertifaiDatasetSource.csv(encoded_data_file))\n",
+    "scan.add_dataset(eval_dataset)\n",
+    "\n",
+    "# Add the explanation dataset - we're only going to bother extracting a small number here\n",
+    "expl_dataset = CertifaiDataset('explanation',\n",
+    "                               CertifaiDatasetSource.dataframe(df[:10]))\n",
+    "scan.add_dataset(expl_dataset)\n",
+    "\n",
+    "# Setup an evaluation for fairness on the above dataset using the model\n",
+    "# We'll look at disparity between groups defined by status, and group it into\n",
+    "# two buckets (male and female) to get at gender bias\n",
+    "female_statuses = [s for s in original_df['status'].unique() if 'female' in s]\n",
+    "male_statuses = [s for s in original_df['status'].unique() if 'female' not in s]\n",
+    "status_grouping = CertifaiGroupingFeature('status',\n",
+    "                                       buckets=[\n",
+    "                                           CertifaiGroupingBucket('female', values=female_statuses),\n",
+    "                                           CertifaiGroupingBucket('male', values=male_statuses)\n",
+    "                                       ])\n",
+    "scan.add_fairness_grouping_feature(status_grouping)\n",
+    "\n",
+    "# Let's look at an additional fairness metric along with counterfactual fairness\n",
+    "scan.add_fairness_metric('demographic parity')\n",
+    "\n",
+    "scan.add_evaluation_type('explanation')\n",
+    "scan.add_evaluation_type('fairness')\n",
+    "scan.evaluation_dataset_id = 'evaluation'\n",
+    "scan.explanation_dataset_id = 'explanation'\n",
+    "scan.hyper_parameter_overrides = {\n",
+    "    'num_counterfactuals': 4\n",
+    "}\n",
+    "\n",
+    "# Because some fields are target-encoded we need to define them in the schema so that\n",
+    "# Certifai knows the value mappings\n",
+    "schema = CertifaiDataSchema()\n",
+    "for feature, value_mappings in target_mappings.items():\n",
+    "    values = []\n",
+    "    encodings = []\n",
+    "    for value, encoding in value_mappings.items():\n",
+    "        values.append(value)\n",
+    "        encodings.append(encoding)\n",
+    "    schema.add_feature(name=feature, data_type=CertifaiFeatureDataType.categorical(values=values,\n",
+    "                                                                                   target_encodings=encodings))\n",
```

Comment:
so the `values` here is in the un-encoded space like `male-single` etc for status and the encodings are the encoded ordered mappings like 1.6 etc . Is the second loop for clarity like we can also use directly `values=list(value_mappings.keys())` and similarly `target_encodings=list(value_mappings.values())`.

Reply:
I did it this way because I don't have complete confidence in order-of-iteration guarantees for dictionaries when using `values` and `keys` (they would have to match)